src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFeatureVariantAssigner.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options;
+using Microsoft.FeatureManagement.FeatureFilters;
+using Microsoft.FeatureManagement.Targeting;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.FeatureManagement
+{
+    /// <summary>
+    /// A feature variant assigner that can be used to assign a variant based on targeted audiences.
+    /// </summary>
+    [AssignerAlias(Alias)]
+    public class ContextualTargetingFeatureVariantAssigner : IContextualFeatureVariantAssigner<ITargetingContext>
+    {
+        private const string Alias = "Microsoft.Targeting";
+        private readonly TargetingEvaluationOptions _options;
+
+        /// <summary>
+        /// Creates a targeting contextual feature filter.
+        /// </summary>
+        /// <param name="options">Options controlling the behavior of the targeting evaluation performed by the filter.</param>
+        public ContextualTargetingFeatureVariantAssigner(IOptions<TargetingEvaluationOptions> options)
+        {
+            _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        /// <summary>
+        /// Assigns one of the variants configured for a feature based off the provided targeting context.
+        /// </summary>
+        /// <param name="variantAssignmentContext">Contextual information available for use during the assignment process.</param>
+        /// <param name="targetingContext">The targeting context used to determine which variant should be assigned.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns></returns>
+        public ValueTask<FeatureVariant> AssignVariantAsync(FeatureVariantAssignmentContext variantAssignmentContext, ITargetingContext targetingContext, CancellationToken cancellationToken)
+        {
+            if (variantAssignmentContext == null)
+            {
+                throw new ArgumentNullException(nameof(variantAssignmentContext));
+            }
+
+            if (targetingContext == null)
+            {
+                throw new ArgumentNullException(nameof(targetingContext));
+            }
+
+            FeatureDefinition featureDefinition = variantAssignmentContext.FeatureDefinition;
+
+            if (featureDefinition == null)
+            {
+                throw new ArgumentNullException(nameof(variantAssignmentContext.FeatureDefinition));
+            }
+
+            if (featureDefinition.Variants == null)
+            {
+                throw new ArgumentNullException(nameof(featureDefinition.Variants));
+            }
```

Comment:
From our discussion on the other PR, should these be `ArgumentException`?

Reply:
Yes, they should.